### PR TITLE
Adding delegated enter/leave events derived from over/move events.

### LINKED ISSUE #239: mouseenter, mouseleave events don't bubble in 3.x
In 2.3 mouseenter and mouseleave bubbled, allowing delegation of these events from a component to it's children:
http://jsbin.com/mevapedehe/1/edit?html,css,console,output

This was due to jQuery mouseenter and mouseleave bubbling while the DOM equivalents do not.

In 3.x the mouseenter events do not bubble, and delegated mouseenter events are not called:
http://jsbin.com/taleliceci/1/edit?html,css,console,output

We need to modify our delegated mouseenter and mouseleave events so they bubble, correcting this inconsistency between 2.3 and 3.x.

We should also confirm that these are the only jQuery events that differ in bubbling from their DOM equivalents.

### PR DESCRIPTION
**Pull request by nlundquist**
Fixes #239.

### CODE REVIEW

File: dom/events/delegate/derive-enter-leave.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  },
+  _addDelegateListener = domEvents.addDelegateListener,
+  _removeDelegateListener = domEvents.removeDelegateListener;
+
+
+domEvents.addDelegateListener = function(eventType, selector, handler) {
+  if (eventMap[eventType] !== undefined) {
+    var origHandler = handler;
+
+    this._derivedHandlers = this._derivedHandlers || {};
```
Comment (nlundquist) [inline dom/events/delegate/derive-enter-leave.js]: Should I keep the reference to the modified handler elsewhere? Use can-util/dom/data instead?

File: dom/events/delegate/derive-enter-leave.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    eventType = eventMap[eventType];
+    handler = function(event) {
+      var target = this,
+        related = event.relatedTarget;
+
+      // For mouseenter/leave call the handler if related is outside the target.
+      // No relatedTarget if the mouse left/entered the browser window
+      if ( !related || (related !== target && !target.contains(related)) ) {
+        return origHandler.call(this, event);
```
Comment (nlundquist) [inline dom/events/delegate/derive-enter-leave.js]: Should I create a PointerEvent or MouseEvent with the event type expected by the delegate handler? Currently this passes an event with the mapped type (ex. passes mouseover event for delegated mouseenter).

I'd just modify the event type on the event, but it's a read only property.

**Comment by nlundquist** [comment]
After this is merged should can-event be modifed to include this? Or are we going to expect users to include this file if they want the functionality?

**Comment by matthewp** [comment]
We talked about putting these types of things into their own packages and having users import them. cc @justinbmeyer

**Comment by justinbmeyer** [comment]
@matthewp IMO, this shouldn't be in another package.  Making "standard" events bubble (like focus/blur) should probably be part of any core event-delegation library we provide.   If we make a `can-dom-event` or something that includes `addDelegateListener`, I'd expect this to be part of it.

File: dom/events/delegate/delegate-test.js [hunk 3]
```diff
@@ -5,6 +5,7 @@ var domEvents = require('can-util/dom/events/');
 var domDispatch = require('can-util/dom/dispatch/');
 var domData = require('can-util/dom/data/');
 require('can-util/dom/events/delegate/');
+require('can-util/dom/events/delegate/derive-enter-leave');
```
Comment (justinbmeyer) [inline dom/events/delegate/delegate-test.js]: why derive as the name?

File: dom/events/delegate/derive-enter-leave.js (hunk 1 above)
Comment (justinbmeyer) [reply dom/events/delegate/derive-enter-leave.js]: yeah, the element isn't a very good place.  I'd checkout the `singleReference` stuff.  You can put the delegating handler on `handler`

File: dom/events/delegate/derive-enter-leave.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      }
+    };
+    this._derivedHandlers[origHandler] = handler;
```
Comment (justinbmeyer) [inline dom/events/delegate/derive-enter-leave.js]: is this going to map to the same thing?  function.tostring()?

**Review by justinbmeyer — CHANGES_REQUESTED** [review]

File: dom/events/delegate/delegate-test.js (hunk 3 above)
Comment (nlundquist) [reply dom/events/delegate/delegate-test.js]: this code derives enter and leave events from other events. i'm open to suggestions

**Comment by justinbmeyer** [comment]
I think enter-leave is enough

File: dom/events/delegate/enter-leave.js [hunk 5]
```diff
@@ -41,7 +62,7 @@ domEvents.addDelegateListener = function(eventType, selector, handler) {
 domEvents.removeDelegateListener = function(eventType, selector, handler) {
   if (eventMap[eventType] !== undefined) {
     eventType = eventMap[eventType];
-    handler = this._derivedHandlers[handler];
+    handler = singleRef.getAndDelete(handler, cid(this)+eventType);
```
Comment (justinbmeyer) [inline dom/events/delegate/enter-leave.js]: yay!

File: dom/events/delegate/enter-leave.js [hunk 6]
```diff
@@ -1,6 +1,8 @@
 'use strict';
 
-var domEvents = require("../events");
+var domEvents = require("../events"),
```
Comment (justinbmeyer) [inline dom/events/delegate/enter-leave.js]: make sure tabs, not spaces are used.  Also, if we did include some of jQuery's code here, make sure to add a comment referencing it and their copyright.

**Review by justinbmeyer — CHANGES_REQUESTED** [review]
add jQuery's copyright if we used some of their source.

File: dom/events/delegate/enter-leave.js (hunk 6 above)
Comment (nlundquist) [reply dom/events/delegate/enter-leave.js]: there's maybe one line that still resembles their code but I'll add the copyright just to be safe

**Comment by nlundquist** [comment]
Only a line or two still resemble JQ but I've added the copyright just to be safe.

**Comment by nlundquist** [comment]
@justinbmeyer once I have a final go ahead I'll merge this, make a minor version release and update can-event to use the new version and require 'enter-leave.js'

**Pull request merged**